ext/dom/parentnode.c

```diff
@@ -240,8 +240,13 @@ static void dom_fragment_assign_parent_node(xmlNodePtr parentNode, xmlNodePtr fr
 
 static zend_result dom_sanity_check_node_list_for_insertion(php_libxml_ref_obj *document, xmlNodePtr parentNode, zval *nodes, int nodesc)
 {
-	if (document == NULL) {
-		php_dom_throw_error(HIERARCHY_REQUEST_ERR, 1);
+	if (UNEXPECTED(document == NULL)) {
+		php_dom_throw_error(HIERARCHY_REQUEST_ERR, 1 /* no document, so be strict */);
```

Comment:
I'm confused, is it really a hierarchy error if we don't have a document?

Reply:
Couldn't find this condition in spec, it was added in https://github.com/php/php-src/pull/5990 to fix https://bugs.php.net/bug.php?id=79968.
However, testing it with a custom element (*) in JS seems to work without an exception:

```js
class Foo extends HTMLElement{}
customElements.define("my-foo", Foo)
foo = new Foo;
foo.append(new Foo);
(new XMLSerializer).serializeToString(foo);
```

(*) custom element, because in JS you can't instantiate an element directly without a document unless you create a custom one.

And you know what's weird? When I remove this check I expected the original bug I linked above to appear again. But nope, no crash :thinking: 
I guess the actual segfault back then happened not because of the missing document, but because of the context node being NULL. However, in current day's code the context node is NULL-checked.
Also calling `dom_get_strict_error` with a NULL argument is fine, because `dom_get_strict_error` will return the default values in that case.

I have pushed a commit to remove that condition and amend the test (again).

Reply:
I've added an additional test for this with elements.

Reply:
Ah so there was a possible crash after all, but I fixed it by reading the document property from the node instead of from the context. The context may be NULL, but the node can't.